File: python_version/linked_lists/linked_list_find/linked_list_find.py
# linked list find
# Write a function, linkedListFind, that takes in a pointer to the head of a linked list and a target value.
#  The function should return a boolean indicating whether or not the linked list contains the target.


class Node:
  def __init__(self, val):
    self.val = val
    self.next = None

def linked_list_find(head, target):
    curr = head
    while curr != None:
        if curr.val == target:
            return True
        curr=curr.next
    return False

def linked_list_find_recur(head, target):
    curr = head
    return _linked_list_find_recur(curr,target)
def _linked_list_find_recur(curr, target):
    if curr == None:
        return False
    if curr.val == target:
        return True
    return _linked_list_find_recur(curr.next,target)




a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")

a.next = b
b.next = c
c.next = d

# a -> b -> c -> d
print("iterative")
print(linked_list_find(a, "c")) # True
print("recursive")
print(linked_list_find_recur(a, "c")) #True

a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")

a.next = b
b.next = c
c.next = d

# a -> b -> c -> d
print("iterative")
print(linked_list_find(a, "d"))# True
print("recursive")
print(linked_list_find_recur(a, "d")) # True

a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")

a.next = b
b.next = c
c.next = d

# a -> b -> c -> d
print("iterative")
print(linked_list_find(a, "q"))# False
print("recursive")
print(linked_list_find_recur(a, "q"))# False


node1 = Node("jason")
node2 = Node("leneli")

node1.next = node2

# jason -> leneli
print("iterative")
print(linked_list_find(node1, "jason")) # True
print("recursive")
print(linked_list_find_recur(node1, "jason")) # True

node1 = Node(42)

# 42
print("iterative")
print(linked_list_find(node1, 42)) # True
print("recursive")
print(linked_list_find_recur(node1, 42)) # True

node1 = Node(42)

# 42
print("iterative")
print(linked_list_find(node1, 100)) # False
print("recursive")
print(linked_list_find_recur(node1, 100)) # False

File: cpp_version/linked_lists/linked_list_find/linked_list_find.cpp
// linked list find
// Write a function, linkedListFind, that takes in a pointer to the head of a linked list and a target value.
//  The function should return a boolean indicating whether or not the linked list contains the target.
#include <cmath>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

class Node {
public:
    std::string value = "";
    Node *next = nullptr;
    Node(std::string initialValue) {
        value = initialValue;
        next = nullptr;
    }
};

bool linkedListFind(Node *, std::string);
bool linkedListFindRecur(Node *, std::string);
bool _linkedListFindRecur(Node *, std::string);

bool linkedListFind(Node *head, std::string target) {
    // todo


    return false;
}

bool linkedListFindRecur(Node *head, std::string target) {
    Node *curr = head;
    return _linkedListFindRecur(curr, target);
}
bool _linkedListFindRecur(Node *curr, std::string target) {
        return false;

}

int main() {

    Node a("a");
    Node b("b");
    Node c("c");
    Node d("d");

    a.next = &b;
    b.next = &c;
    c.next = &d;

    // a -> b -> c -> d

    std::cout << "iterative method" << std::endl;
    std::cout << std::boolalpha << linkedListFind(&a, "c") << std::endl; // 1 (true)
    std::cout << "recursive method" << std::endl;
    std::cout << linkedListFindRecur(&a, "c") << std::endl; // 1 (true)

    // Node a("a");
    // Node b("b");
    // Node c("c");
    // Node d("d");

    // a.next = &b;
    // b.next = &c;
    // c.next = &d;

    // a -> b -> c -> d
    std::cout << "iterative method" << std::endl;
    std::cout << linkedListFind(&a, "d") << std::endl; // 1 (true)
    std::cout << "recursive method" << std::endl;
    std::cout << linkedListFindRecur(&a, "d") << std::endl; // 1 (true)

    // Node a("a");
    // Node b("b");
    // Node c("c");
    // Node d("d");

    // a.next = &b;
    // b.next = &c;
    // c.next = &d;

    // a -> b -> c -> d
    std::cout << "iterative method" << std::endl;
    std::cout << linkedListFind(&a, "q") << std::endl; // 0 (false)
    std::cout << "recursive method" << std::endl;
    std::cout << linkedListFindRecur(&a, "q") << std::endl; // 0 (false)

    Node node1("jason");
    Node node2("leneli");

    node1.next = &node2;

    // jason -> leneli
    std::cout << "iterative method" << std::endl;
    std::cout << linkedListFind(&node1, "jason") << std::endl; // 1 (true)
    std::cout << "recursive method" << std::endl;
    std::cout << linkedListFindRecur(&node1, "jason") << std::endl; // 1 (true)

    Node node3("potato");

    // potato
    std::cout << "iterative method" << std::endl;
    std::cout << linkedListFind(&node3, "potato") << std::endl; // 1 (true)
    std::cout << "recursive method" << std::endl;
    std::cout << linkedListFindRecur(&node3, "potato") << std::endl; // 1 (true)

    // Node node1("potato");

    // potato
    std::cout << "iterative method" << std::endl;
    std::cout << linkedListFind(&node3, "apple") << std::endl; // 0 (false)
    std::cout << "recursive method" << std::endl;
    std::cout << linkedListFindRecur(&node3, "apple") << std::endl; // 0 (false)

    return 0;
}